core/src/retransmit_stage.rs

```diff
@@ -330,19 +362,22 @@ fn retransmit(
                 repair_total += 1;
                 continue;
             }
-
-            if check_if_already_received(packet, shreds_received) {
-                continue;
-            }
-
+            let shred_slot = match check_if_already_received(packet, shreds_received) {
+                Some(slot) => slot,
+                None => continue,
+            };
             let mut compute_turbine_peers = Measure::start("turbine_start");
-            let (my_index, shuffled_stakes_and_index) = ClusterInfo::shuffle_peers_and_index(
+            let (my_index, mut shuffled_stakes_and_index) = ClusterInfo::shuffle_peers_and_index(
                 &my_id,
                 &r_epoch_stakes_cache.peers,
                 &r_epoch_stakes_cache.stakes_and_index,
                 packet.meta.seed,
             );
             peers_len = cmp::max(peers_len, shuffled_stakes_and_index.len());
+            // Until the patch is activated, do the old buggy thing.
+            if !enable_turbine_retransmit_peers_patch(shred_slot, bank_forks, &mut banks_cache) {
```

Comment:
How is banks_cached pruned of old banks?

Reply:
It is a local variable to `fn retransmit`, so it is dropped at function exit.
And its size is bounded by `MAX_PACKET_BATCH_SIZE`.

Reply:
I see. I'm not sure the batches will be all that big, the cache may not be very effective at this level.

Reply:
I switched to an `LruCache` preserved between calls to `fn retransmit(...)`.

One question is that, what to do if `bank_forks.get(shred_slot)` returns `None`:
* I am using `bank_forks.root_bank()` as the fallback. Would that make sense?
* Should it cache that fallback bank for the `shred_slot`?
I chose not to cache that, because following lookups will still use the fallback bank even if `bank_forks.get(shred_slot)` no longer is `None`. This is probably safer approach, but can potentially cause more locking on `bank_forks`.

Reply:
I think you would want to find a bank with the same epoch as the `shred.slot`. `working_bank` would probably be more likely, because `root_bank` will trail the epoch start by some amount which is takes to root the banks coming into the epoch.

But to be accurate, you could save a bank polled from `working_bank` when the epoch changes and then have both epoch N, N+1 banks to use to compute it.

Reply:
I am not sure I understand your suggestion. Can you please clarify?

Would it make sense to go through all banks in bank_forks:
https://github.com/solana-labs/solana/blob/d63b2baf0/runtime/src/bank_forks.rs#L46
to find one with the same `epoch` as `shred.slot`?
And, if so, then
* Which one should it be used if there are multiple matches?
* What to do if there are no matches?

Reply:
> Would it make sense to go through all banks in bank_forks?

Yea, that's an option.

> Which one should it used if there are multiple matches?

It shouldn't matter, any bank in the epoch should give the same answer, first one it finds.

> What to do if there are no matches?

Using `root_bank` should be ok then.